File: internal/content/testdata/bench/rust/nc-find-median-stream.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

fn median_stream_ops(operations: Vec<Vec<J>>) -> Vec<J> {
    // lo: max-heap (lower half), hi: min-heap (upper half).
    let mut lo: BinaryHeap<i64> = BinaryHeap::new();
    let mut hi: BinaryHeap<Reverse<i64>> = BinaryHeap::new();
    let mut out: Vec<J> = Vec::new();

    for op in &operations {
        let name = match &op[0] {
            J::Str(s) => s.as_str(),
            _ => "",
        };
        if name == "addNum" {
            let v = if let J::Int(x) = &op[1] { *x } else { 0 };
            lo.push(v);
            // move top of lo to hi
            let t = lo.pop().unwrap();
            hi.push(Reverse(t));
            if hi.len() > lo.len() {
                let Reverse(t2) = hi.pop().unwrap();
                lo.push(t2);
            }
            out.push(J::Null);
        } else {
            // findMedian
            if lo.len() > hi.len() {
                out.push(J::Int(*lo.peek().unwrap()));
            } else {
                let a = *lo.peek().unwrap();
                let b = hi.peek().unwrap().0;
                let sum = a + b;
                if sum % 2 == 0 {
                    out.push(J::Int(sum / 2));
                } else {
                    out.push(J::Flt(sum as f64 / 2.0));
                }
            }
        }
    }
    out
}

File: internal/content/testdata/bench/go/nc-find-median-stream.go
import "container/heap"

type maxHeap []int
func (h maxHeap) Len() int           { return len(h) }
func (h maxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h maxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *maxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *maxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

type minHeap []int
func (h minHeap) Len() int           { return len(h) }
func (h minHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h minHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *minHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

type MedianFinder struct {
	lo *maxHeap // lower half, max-heap
	hi *minHeap // upper half, min-heap
}

func newMedianFinder() *MedianFinder {
	lo := &maxHeap{}
	hi := &minHeap{}
	heap.Init(lo)
	heap.Init(hi)
	return &MedianFinder{lo: lo, hi: hi}
}

func (m *MedianFinder) addNum(v int) {
	heap.Push(m.lo, v)
	heap.Push(m.hi, heap.Pop(m.lo).(int))
	if m.hi.Len() > m.lo.Len() {
		heap.Push(m.lo, heap.Pop(m.hi).(int))
	}
}

func (m *MedianFinder) findMedian() float64 {
	if m.lo.Len() > m.hi.Len() {
		return float64((*m.lo)[0])
	}
	return float64((*m.lo)[0]+(*m.hi)[0]) / 2.0
}

func median_stream_ops(operations [][]interface{}) []interface{} {
	mf := newMedianFinder()
	out := make([]interface{}, len(operations))
	for i, op := range operations {
		switch op[0].(string) {
		case "addNum":
			mf.addNum(op[1].(int))
			out[i] = nil
		default: // findMedian
			out[i] = mf.findMedian()
		}
	}
	return out
}
